pallets/solo-to-para/src/lib.rs

```diff
@@ -0,0 +1,178 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Decode, Encode};
+use cumulus_pallet_parachain_system as parachain_system;
+use frame_support::{dispatch::DispatchResult, pallet_prelude::*, weights::DispatchInfo};
+use frame_system::pallet_prelude::*;
+pub use pallet::*;
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{DispatchInfoOf, Dispatchable, SignedExtension},
+	transaction_validity::{
+		InvalidTransaction, TransactionLongevity, TransactionPriority, TransactionValidity,
+		TransactionValidityError, ValidTransaction,
+	},
+};
+use sp_std::{prelude::*, vec::Vec};
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config:
+		frame_system::Config + cumulus_pallet_parachain_system::Config + pallet_sudo::Config
+	{
+		type Event: From<Event> + IsType<<Self as frame_system::Config>::Event>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	/// In case of a scheduled migration, this storage field contains the custom head data to be applied.
+	#[pallet::storage]
+	pub(super) type PendingCustomValidationHeadData<T: Config> =
+		StorageValue<_, Vec<u8>, OptionQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event {
+		MigrationScheduled,
+		/// The custom validation head data has been scheduled to apply.
+		CustomValidationHeadDataStored,
+		/// The custom validation head data was applied as of the contained relay chain block number.
+		CustomValidationHeadDataApplied, //(RelayChainBlockNumber),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// CustomHeadData is not stored in storage.
+		NoCustomHeadData,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(0)]
+		pub fn schedule_migration(
+			origin: OriginFor<T>,
+			code: Vec<u8>,
+			head_data: Vec<u8>,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+
+			parachain_system::Pallet::<T>::set_code_impl(code)?;
+			Self::store_pending_custom_validation_head_data(head_data);
+
+			Self::deposit_event(Event::MigrationScheduled);
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Set a custom head data that should only be applied when upgradeGoAheadSignal from
+		/// the Relay Chain is GoAhead
+		fn store_pending_custom_validation_head_data(head_data: Vec<u8>) {
+			PendingCustomValidationHeadData::<T>::put(head_data);
+			Self::deposit_event(Event::CustomValidationHeadDataStored);
```

Comment:
Second event that is being send as part of `schedule_migration`

Reply:
I removed one of the events. Now we emit only `CustomValidationHeadDataStored`